INTMDB-364 Add support for serverless private endpoints (AWS + Azure) (#913)
mongodb/terraform-provider-mongodbatlas: website/docs/r/privatelink_endpoint_serverless.html.markdown

---
layout: "mongodbatlas"
page_title: "MongoDB Atlas: privatelink_endpoint_serverless"
sidebar_current: "docs-mongodbatlas-datasource-privatelink-endpoint-serverless"
description: |-

> Describes a Serverless PrivateLink endpoint

Reviewer: ```suggestion
Describes a Serverless PrivateLink Endpoint
```

Author: Corrected

---

# Data Source: privatelink_endpoint_serverless

> `privatelink_endpoint_serverless` Provides a Serverless PrivateLink endpoint resource.

Reviewer: ```suggestion
`privatelink_endpoint_serverless` Provides a Serverless PrivateLink Endpoint resource.
```

Author: Corrected

> `privatelink_endpoint_serverless` Provides a Serverless PrivateLink Endpoint resource.

Reviewer 2: It would be nice to start including more help on these when two resources are required.  Can we add something like: 
```suggestion
`privatelink_endpoint_serverless` Provides a Serverless PrivateLink Endpoint resource.  This is the first of two resources required to configure PrivateLink for Serverless, the second is [mongodbatlas_privatelink_endpoint_service_serverless](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/privatelink_endpoint_service_serverless).  
```

-> **NOTE:** Groups and projects are synonymous terms. You may find group_id in the official documentation.

## Example Usage

### Basic
```terraform

resource "mongodbatlas_privatelink_endpoint_serverless" "test" {
	project_id   = "<PROJECT_ID>"
	instance_name = mongodbatlas_serverless_instance.test.name
	provider_name = "AWS"
}
	  
resource "mongodbatlas_serverless_instance" "test" {
	project_id   = "<PROJECT_ID>"
	name         = "test-db"
	provider_settings_backing_provider_name = "AWS"
	provider_settings_provider_name = "SERVERLESS"
	provider_settings_region_name = "US_EAST_1"
	continuous_backup_enabled = true
}
```

## Argument Reference

* `project_id` - (Required) Unique 24-digit hexadecimal string that identifies the project.

> * `instance_name` - (Required) Serverless instance name to attach private

Reviewer: ```suggestion
* `instance_name` - (Required) Human-readable label that identifies the serverless instance associated with the tenant endpoint
```

Author: Corrected

> * `instance_name` - (Required) Human-readable label that identifies the serverless instance associated with the tenant endpoint

Reviewer 2: ```suggestion
* `instance_name` - (Required) Human-readable label that identifies the serverless instance 
```

> * `provider_name` - (Required) Cloud provider name AWS currently supported

Reviewer: ```suggestion
* `provider_name` - (Required) Cloud provider name; AWS is currently supported
```

Author: Corrected

## Attributes Reference

In addition to all arguments above, the following attributes are exported:
* `endpoint_id` - Unique 22-character alphanumeric string that identifies the private endpoint. Atlas supports AWS private endpoints using the [|aws| PrivateLink](https://aws.amazon.com/privatelink/) feature.

> * `endpoint_service_name` - Unique string that identifies the PrivateLink endpoint service. MongoDB Cloud returns null while it creates the endpoint service.

Reviewer 2: Do we really need to mention the null return here?  Will the Terraform user ever see that?  
```suggestion
* `endpoint_service_name` - Unique string that identifies the PrivateLink endpoint service. 
```

Reviewer: ![image](https://user-images.githubusercontent.com/47370943/201209202-d28d8e76-d2e8-4b6d-a7bc-fe2b88803ddb.png)
that's the definition we have in new docs: https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Serverless-Private-Endpoints/operation/createOnePrivateEndpointForOneServerlessInstance

Reviewer: agree doesn't make sense for external user using terraform, we should delete second sentence

* `cloud_provider_endpoint_id` - Unique string that identifies the private endpoint's network interface.
* `comment` - Human-readable string to associate with this private endpoint.
* `status` - Human-readable label that indicates the current operating status of the private endpoint. Values include: RESERVATION_REQUESTED, RESERVED, INITIATING, AVAILABLE, FAILED, DELETING.

> For more information see: [MongoDB Atlas API - Serverless Private Endpoints](https://www.mongodb.com/docs/atlas/reference/api/serverless-private-endpoints-get-one/)  and [MongoDB Atlas API - Online Archive](https://docs.atlas.mongodb.com/reference/api/online-archive/) Documentation.

Reviewer 2: ```suggestion
For more information see: [MongoDB Atlas API - Serverless Private Endpoints](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Serverless-Private-Endpoints/operation/createOnePrivateEndpointForOneServerlessInstance/).
```

---
layout: "mongodbatlas"
page_title: "MongoDB Atlas: privatelink_endpoint_serverless"
sidebar_current: "docs-mongodbatlas-datasource-privatelink-endpoint-serverless"
description: |-
Describes a Serverless PrivateLink Endpoint
---

> # Data Source: privatelink_endpoint_serverless

Reviewer: ```suggestion
# Resource: privatelink_endpoint_serverless
```

`privatelink_endpoint_serverless` Provides a Serverless PrivateLink Endpoint resource.
This is the first of two resources required to configure PrivateLink for Serverless, the second is [mongodbatlas_privatelink_endpoint_service_serverless](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/privatelink_endpoint_service_serverless).

> -> **NOTE:** Groups and projects are synonymous terms. You may find group_id in the official documentation.

Reviewer: ```suggestion
-> **NOTE:** Groups and projects are synonymous terms. You may find group_id in the official documentation.

-> **NOTE:** Feature unavailable in Free and Shared-Tier Clusters
This feature is not available for M0 free clusters, M2, and M5 clusters. To learn more about which features are unavailable, see [Atlas M0 (Free Cluster), M2, and M5 Limitations.](https://www.mongodb.com/docs/atlas/reference/free-shared-limitations/#std-label-atlas-free-tier)
```

Reviewer 2: @Zuhairahmed I actually don't think this note makes sense here.   Why would you call out that something with serverless in the name is not available for free/shared tier?  I get it maybe for general privatelink but not this one.

Reviewer: mainly since we call this out our Atlas Admin API/UI Set Up page for this resource: https://www.mongodb.com/docs/atlas/security-serverless-private-endpoint/

Reviewer: mainly since we do this Set Up page for Atlas Admin API/UI: https://www.mongodb.com/docs/atlas/security-serverless-private-endpoint/

Reviewer: @martinstibbe delete this one

## Example Usage

### Basic
```terraform

resource "mongodbatlas_privatelink_endpoint_serverless" "test" {
	project_id   = "<PROJECT_ID>"
	instance_name = mongodbatlas_serverless_instance.test.name
	provider_name = "AWS"
}
	  
resource "mongodbatlas_serverless_instance" "test" {
	project_id   = "<PROJECT_ID>"
	name         = "test-db"
	provider_settings_backing_provider_name = "AWS"
	provider_settings_provider_name = "SERVERLESS"
	provider_settings_region_name = "US_EAST_1"
	continuous_backup_enabled = true
}
```

## Argument Reference

* `project_id` - (Required) Unique 24-digit hexadecimal string that identifies the project.
* `instance_name` - (Required) Human-readable label that identifies the serverless instance.
* `provider_name` - (Required) Cloud provider name; AWS is currently supported

## Attributes Reference

In addition to all arguments above, the following attributes are exported:

> * `endpoint_id` - Unique 22-character alphanumeric string that identifies the private endpoint. Atlas supports AWS private endpoints using the [|aws| PrivateLink](https://aws.amazon.com/privatelink/) feature.

Reviewer: ```suggestion
* `endpoint_id` - Unique 22-character alphanumeric string that identifies the private endpoint. Atlas supports AWS private endpoints using the [AWS PrivateLink](https://aws.amazon.com/privatelink/) feature.
```

* `endpoint_service_name` - Unique string that identifies the PrivateLink endpoint service.
* `private_link_service_resource_id` - Root-relative path that identifies the Azure Private Link Service that MongoDB Cloud manages.
* `cloud_provider_endpoint_id` - Unique string that identifies the private endpoint's network interface.
* `comment` - Human-readable string to associate with this private endpoint.
* `status` - Human-readable label that indicates the current operating status of the private endpoint. Values include: RESERVATION_REQUESTED, RESERVED, INITIATING, AVAILABLE, FAILED, DELETING.

> For more information see: [MongoDB Atlas API - Serverless Private Endpoints](https://www.mongodb.com/docs/atlas/reference/api/serverless-private-endpoints-get-one/).

Reviewer: Missing IMPORT section

---
layout: "mongodbatlas"
page_title: "MongoDB Atlas: privatelink_endpoint_serverless"
sidebar_current: "docs-mongodbatlas-datasource-privatelink-endpoint-serverless"
description: |-
Describes a Serverless PrivateLink Endpoint
---

# Resource: privatelink_endpoint_serverless

`privatelink_endpoint_serverless` Provides a Serverless PrivateLink Endpoint resource.
This is the first of two resources required to configure PrivateLink for Serverless, the second is [mongodbatlas_privatelink_endpoint_service_serverless](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/privatelink_endpoint_service_serverless).

-> **NOTE:** Groups and projects are synonymous terms. You may find group_id in the official documentation.
-> **NOTE:** Feature unavailable in Free and Shared-Tier Clusters
This feature is not available for M0 free clusters, M2, and M5 clusters. To learn more about which features are unavailable, see [Atlas M0 (Free Cluster), M2, and M5 Limitations.](https://www.mongodb.com/docs/atlas/reference/free-shared-limitations/#std-label-atlas-free-tier)
## Example Usage

> ### Basic

Reviewer 2: More a nit but if there's a "Basic" I expect an advanced option.  But this doesn't really have an advanced option.  Not blocking but probably better to remove it or rename it something like "AWS Example".

```terraform

resource "mongodbatlas_privatelink_endpoint_serverless" "test" {
	project_id   = "<PROJECT_ID>"
	instance_name = mongodbatlas_serverless_instance.test.name
	provider_name = "AWS"
}
	  
resource "mongodbatlas_serverless_instance" "test" {
	project_id   = "<PROJECT_ID>"
	name         = "test-db"
	provider_settings_backing_provider_name = "AWS"
	provider_settings_provider_name = "SERVERLESS"
	provider_settings_region_name = "US_EAST_1"
	continuous_backup_enabled = true
}
```

## Argument Reference

* `project_id` - (Required) Unique 24-digit hexadecimal string that identifies the project.
* `instance_name` - (Required) Human-readable label that identifies the serverless instance.
* `provider_name` - (Required) Cloud provider name; AWS is currently supported

> 

Reviewer 2: Since you added timeouts don't we need something like https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/privatelink_endpoint_service#timeouts?

## Attributes Reference

In addition to all arguments above, the following attributes are exported:

> * `endpoint_id` - Unique 22-character alphanumeric string that identifies the private endpoint. Atlas supports AWS private endpoints using the [AWS PrivateLink](https://aws.amazon.com/privatelink/) feature.

Reviewer 2: Since both AWS and Azure just remove the AWS comment.  Also id returned is 24 - copied this from the original docs: https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Serverless-Private-Endpoints/operation/returnAllPrivateEndpointsForOneServerlessInstance
```suggestion
* `endpoint_id` - Unique 24-hexadecimal digit string that identifies the private endpoint.
```
